File: src/data/courses/dsa/binary-search-trees/views-vertical-problems/vertical-sum/code/java/VerticalSum.java
import java.util.*;
class Solution {
    public List<Integer> verticalSum(TreeNode root) {
        TreeMap<Integer,Integer> sums = new TreeMap<>();
        Queue<Object[]> q = new LinkedList<>(); if(root!=null)q.add(new Object[]{root,0});
        while(!q.isEmpty()){Object[]c=q.poll();TreeNode n=(TreeNode)c[0];int hd=(int)c[1];
            sums.merge(hd,n.val,Integer::sum);
            if(n.left!=null)q.add(new Object[]{n.left,hd-1});
            if(n.right!=null)q.add(new Object[]{n.right,hd+1});}
        return new ArrayList<>(sums.values());
    }
}

File: src/data/courses/dsa/binary-search-trees/views-vertical-problems/vertical-sum/code/python/vertical_sum.py
from collections import deque, defaultdict
class Solution:
    def verticalSum(self, root):
        if not root: return []
        sums = defaultdict(int)
        queue = deque([(root, 0)])
        while queue:
            node, hd = queue.popleft()
            sums[hd] += node.val
            if node.left: queue.append((node.left, hd-1))
            if node.right: queue.append((node.right, hd+1))
        return [sums[k] for k in sorted(sums)]
